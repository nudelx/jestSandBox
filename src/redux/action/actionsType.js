export const navButtonClick = 'NAV_BUTTON::CLICK'
