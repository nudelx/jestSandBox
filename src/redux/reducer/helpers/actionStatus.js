 const actionStatus = {
  FETTCHING: 'FETTCHING',
  DONE: 'DONE',
  FAIL: 'FAIL',
}

export default actionStatus
