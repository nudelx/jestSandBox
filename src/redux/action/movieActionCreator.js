import { MovieFetching, MovieFetchDone, MovieFetchError } from "./actionsType"
import MovieAPI from "../../helper/MovieAPI"

export const MovieActionsCreator = {
  startGettingMovie: page => ({ type: MovieFetching, page }),
  getMovieDone: (movie, page) => ({ type: MovieFetchDone, movie, page }),
  getMovieError: error => ({ type: MovieFetchError, error })
}

export const tryGetMovie = movieNum =>
  movieNum && MovieAPI.fetch(`https://swapi.co/api/films/${movieNum}`)

export const fetchingMovieProcess = page => (dispatch, getState) => {
  if (getState().movies[page] && getState().movies[page].status === "DONE") {
    return Promise.resolve(getState().movies[page].data)
  }

  const [_page, movieNum] = page.split("_") // eslint-disable-line
  dispatch(MovieActionsCreator.startGettingMovie(page))
  return tryGetMovie(movieNum)
    .then(data => dispatch(MovieActionsCreator.getMovieDone(data, page)))
    .catch(error => dispatch(MovieActionsCreator.getMovieError(error, page)))
}
