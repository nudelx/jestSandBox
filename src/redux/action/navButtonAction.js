import { navButtonClick } from './actionsType'

const NavButtonActionsCreator = {
  createNavButtonClick: page => ({ type: navButtonClick, page })
}



export default NavButtonActionsCreator
