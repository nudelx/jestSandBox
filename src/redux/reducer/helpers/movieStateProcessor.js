import {
  MovieFetching,
  MovieFetchDone,
  MovieFetchError
} from '../../action/actionsType'
import actionStatus from './actionStatus'

export const defaultState = {}

export const stateProcessor = {
  [MovieFetching]: (state, action) => {
    return {
      ...state,
      [action.page]: { status: actionStatus.FETCHING, data: null }
    }
  },
  [MovieFetchDone]: (state, action) => {
    console.log('action', action)
    return {
      ...state,
      [action.page]: { status: actionStatus.DONE, data: action.movie }
    }
  },
  [MovieFetchError]: (state, action) => {
    return {
      ...state,
      [action.page]: {
        status: actionStatus.FAIL,
        data: null,
        err: action.error
      }
    }
  }
}
