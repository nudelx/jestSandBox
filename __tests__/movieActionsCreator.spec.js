import { stateProcessor } from '../src/redux/reducer/helpers/movieStateProcessor'
import  { MovieFetching, MovieFetchDone, MovieFetchError } from  '../src/redux/action/actionsType'
import enzymeSerializer from 'enzyme-to-json/serializer';
import { MovieActionsCreator,  tryGetMovie }  from '../src/redux/action/movieActionCreator'

const page  = 'page_1'
const movie = { movie: 'test movie '}
const error = 'this is a test error'
describe('it create actions propertly', () => {

  it('MovieFetching', () => {
    expect.addSnapshotSerializer(enzymeSerializer)
    const state = MovieActionsCreator.startGettingMovie(page)
    expect(state).toMatchSnapshot()
  })

  it('getMovieDone', () => {
    expect.addSnapshotSerializer(enzymeSerializer)
    const state = MovieActionsCreator.getMovieDone(movie,page)
    expect(state).toMatchSnapshot()
  })

  it('getMovieError', () => {
    expect.addSnapshotSerializer(enzymeSerializer)
    const state = MovieActionsCreator.getMovieError(error)
    expect(state).toMatchSnapshot()
  })


})


describe('it should try to get a movie', () => {

  it('getMovieError', () => {
    expect.addSnapshotSerializer(enzymeSerializer)
    tryGetMovie(1).then(d => d.json()).then(d => {
      console.log(d)
      expect(1).toBe(1)
    })

  })


})
