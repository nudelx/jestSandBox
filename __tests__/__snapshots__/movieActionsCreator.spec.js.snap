// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it creates actions propertly MovieFetching 1`] = `
Object {
  "page": "page_1",
  "type": "MOVIE::FETCHING",
}
`;

exports[`it creates actions propertly getMovieDone 1`] = `
Object {
  "movie": Object {
    "movie": "test movie ",
  },
  "page": "page_1",
  "type": "MOVIE::FETCHING::DONE",
}
`;

exports[`it creates actions propertly getMovieError 1`] = `
Object {
  "error": "this is a test error",
  "type": "MOVIE::FETCHING::ERROR",
}
`;

exports[`it should try to get a movie get a movie 1 1`] = `
Object {
  "created": "2014-12-10 14:23:31",
  "director": "BB-8",
  "edited": "2015-04-11 09:46:52",
  "episode_id": [Circular],
  "opening_crawl": "this is a test",
  "producer": "Kylo Ren",
  "release_date": "1977-05-25",
  "title": "Hope it Will work",
  "url": "/this/is/url/to/some/movie/",
}
`;
